src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+        /// </summary>
+        private void MoveDown((TElement element, TPriority priority) node, int nodeIndex)
+        {
+            // The node to move down will not actually be swapped every time.
+            // Rather, values on the affected path will be moved up, thus leaving a free spot
+            // for this value to drop in. Similar optimization as in the insertion sort.
+
+            int i;
+            while ((i = this.GetFirstChildIndex(nodeIndex)) < this.nodes.Count)
+            {
+                // Check if the current node (pointed by 'nodeIndex') should really be extracted
+                // first, or maybe one of its children should be extracted earlier.
+                var topChild = this.nodes[i];
+                var childrenIndexesLimit = Math.Min(i + Arity, this.nodes.Count);
+                int topChildIndex = i;
+
+                while (++i < childrenIndexesLimit)
+                {
+                    var child = this.nodes[i];
+                    if (this.Comparer.Compare(child.priority, topChild.priority) < 0)
+                    {
+                        topChild = child;
+                        topChildIndex = i;
+                    }
+                }
+
+                // In case no child needs to be extracted earlier than the current node,
+                // there is nothing more to do - the right spot was found.
+                if (this.Comparer.Compare(node.priority, topChild.priority) <= 0)
+                {
+                    break;
+                }
+
+                // Move the top child up by one node and now investigate the
+                // node that was considered to be the top child (recursive).
+                this.PutAt(topChild, nodeIndex);
+                nodeIndex = topChildIndex;
+            }
+
+            this.PutAt(node, nodeIndex);
+        }
+
+        /// <summary>
+        /// Gets a collection that enumerates the elements of the queue.
+        /// </summary>
+        public UnorderedItemsCollection UnorderedItems { get; }
+
+        public partial class UnorderedItemsCollection : IReadOnlyCollection<(TElement element, TPriority priority)>, ICollection
+        {
+            private readonly IReadOnlyCollection<(TElement element, TPriority priority)> items;
+
+            public UnorderedItemsCollection(IReadOnlyCollection<(TElement element, TPriority priority)> items)
+            {
+                this.items = items;
+            }
+
+            public int Count => this.items.Count;
+            object ICollection.SyncRoot => this;
+            bool ICollection.IsSynchronized => false;
+
+            public void CopyTo(Array array, int index) => throw new NotImplementedException();
+
+            public struct Enumerator : IEnumerator<(TElement element, TPriority priority)>, IEnumerator
+            {
+                private readonly IEnumerator<(TElement element, TPriority priority)> enumerator;
```

Comment:
This will change anyway once you stop using List, but this is going to box the List's enumerator; it should be using the strongly-typed List.Enumerator struct.